core/apple/src/AppleCore.kt

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2017-2023 JetBrains s.r.o. and respective authors and developers.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the LICENCE file.
+ */
+
+package kotlinx.io
+
+import kotlinx.cinterop.*
+import platform.Foundation.NSInputStream
+import platform.Foundation.NSOutputStream
+import platform.Foundation.NSStreamStatusNotOpen
+import platform.darwin.UInt8Var
+
+/**
+ * Returns [RawSink] that writes to an output stream.
+ *
+ * Use [RawSink.buffered] to create a buffered sink from it.
+ *
+ * @sample kotlinx.io.samples.KotlinxIoSamplesApple.outputStreamAsSink
+ */
+public fun NSOutputStream.asSink(): RawSink = OutputStreamSink(this)
+
+private open class OutputStreamSink(
+    private val out: NSOutputStream,
+) : RawSink {
+
+    @OptIn(UnsafeNumber::class)
+    override fun write(source: Buffer, byteCount: Long) {
+        if (out.streamStatus == NSStreamStatusNotOpen) out.open()
```

Comment:
may be we also need to check for closed state here and throw right here?

Reply:
> Interestingly, you can't close an NSOutputStream that's never been opened. So you have to write something before you can close the Sink too.

So what behaviour we will have if we do `NSOutputStream.asSink().close()`? (so when stream is not opened)

Reply:
> So what behaviour we will have if we do NSOutputStream.asSink().close()? (so when stream is not opened)

The stream will remain in the `NSStreamStatusNotOpen` state. If we want to avoid this, we could `open()` and then `close()` the stream if it is in this state. That could end up doing extra work unnecessarily though, depending on the `NSOutputStream` implementation. Not sure which is preferable.

Reply:
I see the issue with how closing is handled now; consider the following case:
```
val source = NSInputStream.inputStreamWithFileAtPath("/dev/random")!!.asSource()
source.close()
assertFailsWith<Throwable> {
        source.readAtMostTo(Buffer(), 10)
}
```
The test will fail because `close` is effectively doing nothing here (the stream is not opened yet) and the next read attempt will open the stream and it won't be considered as closed.

I'm fine with opening the stream on the first read as long as the behavior will be reflected in the class' documentation. But the close should be handled properly - if we closed the source before reading for it then the next read should fail.
The same is true for sinks.

Reply:
So should we `open()` and then `close()` the underlying `NSInputStream` or `NSOutputStream` if closed before read or write? Or should we `open()` it when the `Source` or `Sink` is created in `init`?

Reply:
Personally, I like the idea of opening the stream in the `init` more.

Reply:
I think that makes more sense to me as well, now that we've iterated through the behavior of delaying the opening until first read/write, since essentially we're saying the stream must be opened eventually however you end up using it (assuming valid usage should always close the stream). The only benefit to delaying opening might be to prevent potential resource leakage if a user creates a stream but never reads/writes or closes the stream, which should be unexpected. I'll go ahead and revert to opening on `init`.